oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/resourceserver/web/BearerTokenAuthenticationEntryPoint.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.resourceserver.web;
+
+import java.io.IOException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.oauth2.resourceserver.BearerTokenAuthenticationException;
+import org.springframework.security.oauth2.resourceserver.BearerTokenError;
+import org.springframework.security.web.AuthenticationEntryPoint;
+import org.springframework.util.Assert;
+
+/**
+ * An {@link AuthenticationEntryPoint} implementation used to commence authentication of protected resource requests
+ * using {@link BearerTokenAuthenticationFilter}.
+ * <p>
+ * Uses information provided by {@link BearerTokenError} to set HTTP response status code and populate
+ * {@code WWW-Authenticate} HTTP header.
+ *
+ * @author Vedran Pavic
+ * @since 5.1
+ * @see BearerTokenAuthenticationException
+ * @see BearerTokenError
+ * @see <a href="https://tools.ietf.org/html/rfc6750#section-3" target="_blank">RFC 6750 Section 3: The WWW-Authenticate
+ * Response Header Field</a>
+ */
+public class BearerTokenAuthenticationEntryPoint implements AuthenticationEntryPoint {
+
+	private String realmName;
+
+	@Override
+	public void commence(HttpServletRequest request, HttpServletResponse response,
+			AuthenticationException authException) throws IOException {
+		HttpStatus httpStatus;
+		Map<String, String> authParamAttributes = new LinkedHashMap<>();
+		if (this.realmName != null) {
+			authParamAttributes.put("realm", this.realmName);
+		}
+		if (authException instanceof BearerTokenAuthenticationException) {
+			BearerTokenError error = ((BearerTokenAuthenticationException) authException).getError();
+			httpStatus = error.getHttpStatus();
+			authParamAttributes.put("error", error.getErrorCode());
+			String description = error.getDescription();
+			if (description != null) {
+				authParamAttributes.put("error_description", description);
+			}
+			String uri = error.getUri();
+			if (uri != null) {
+				authParamAttributes.put("error_uri", uri);
+			}
+			String scope = error.getScope();
```

Comment:
What is your thought around descriptions like the following in the spec:

> Values for the "scope" attribute MUST NOT include characters outside the set %x21 / %x23-5B
   / %x5D-7E for representing scope values and %x20 for delimiters
   between scope values.

I think the big one is (").

Reply:
I didn't look too deep in that part for now, will give this some thought.

Reply:
`BearerTokenAuthenticationEntryPointTests` doesn't seem the right place to validate this IMO, as it only retrieves whatever scope value is passed in with the error. Validation should happen in place where scope value (that can end up in error) is configured.

Reply:
Fair. Would you mind creating a corresponding ticket?